### PR TITLE
WIP Dataset

### PR DESCRIPTION
**Pull request by fedorgrab**

### CODE REVIEW

File: src/casp/ml/data/transforms.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            format_string += "\n"
+            format_string += f"    {transform}"
+
+        format_string += "\n)"
+        return format_string
+
+
+class RowWiseNormalization(CASTransform):
+    def __init__(self, C: int):
+        self.C = C
```
Comment (mbabadi) [inline src/casp/ml/data/transforms.py:36]: @fedorgrab please rename `C` to `SC_RNA_NORMALIZATION_PSEUDOCOUNT`.

File: src/casp/ml/services/inference/pca/constants.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import json
+import os
+import typing as t
+
+import dotenv
+
+dotenv.load_dotenv(dotenv_path="casp/ml/services/inference/.env")
```
Comment (kcibul) [inline src/casp/ml/services/inference/pca/constants.py:7]: it feels odd to have two ways of doing this, some via constants some via env vars?

File: src/casp/ml/services/training/pca/utils.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import typing as t
+
+from google.oauth2.service_account import Credentials
+
+from casp.ml.services.training.pca import constants
+
+
+def get_google_service_credentials() -> t.Tuple["Credentials", str]:
```
Comment (kcibul) [inline src/casp/ml/services/training/pca/utils.py:8]: feels like I've seen this method before?

**Review by kcibul — APPROVED** [review]

File: src/casp/ml/running_stats.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            self.running_sums = torch.zeros(tensor.shape, device=tensor.device)
+            self.running_sums_squared = torch.zeros(tensor.shape, device=tensor.device)
+
+        self.running_sums += tensor
+        self.running_sums_squared += torch.square(tensor)
+        self.n += 1
+        return tensor
+
+
+class OnePassMeanVarStd(RunningSums):
```
[the remark this answers is not in the document]
Comment (fedorgrab) [reply src/casp/ml/running_stats.py:33]: I was trying to show that the mean, var, and std are calculated in [one-pass technique](https://en.wikipedia.org/wiki/One-pass_algorithm).

File: src/casp/ml/data/transforms.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+
+class CASTransform(PickleMixin):
+    def __call__(self, tensor: torch.Tensor) -> torch.Tensor:
+        raise NotImplementedError
+
+
+class Compose(CASTransform):
+    """
+    Inspired by `torchvision.transforms.Compose`
```
[the remark this answers is not in the document]
Comment (fedorgrab) [reply src/casp/ml/data/transforms.py:16]: Yeah, we sure can, but it's a part of `torchvision` module which is made for image recognition. I thought it would be confusing to use. That was the only reason.

File: src/casp/ml/services/inference/pca/constants.py (hunk 2 above)
Comment (fedorgrab) [reply src/casp/ml/services/inference/pca/constants.py:7]: I don't think so, as in `.env` I store secret variables like service account credentials. I parse dot env and use it as a secret environment variable. Later on `.env` is supposed to be taken from secret storage and generated by CI. I am happy to see your suggestions tho. But I think if we put everything to env, it wouldn't be that obvious that our settings/constants file has these variables.

File: src/casp/ml/data/transforms.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        for transform in self.transforms:
+            format_string += "\n"
+            format_string += f"    {transform}"
+
+        format_string += "\n)"
+        return format_string
+
+
+class RowWiseNormalization(CASTransform):
+    def __init__(self, sc_rna_normalization_pseudocount: int):
```
[the remark this answers is not in the document]
Comment (fedorgrab) [reply src/casp/ml/data/transforms.py:38]: I believe it is called so because this is a factor that they use in the formula for normalizing scRNA data. Like it's scRNA normalization-specific variable that makes it different from other techniques that are not used in scRNA data. It was Mehrtash's recommendation to call this variable like this. I am sure he can explain the logic behind it better than me.)

File: src/casp/ml/data/transforms.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (fedorgrab) [reply src/casp/ml/data/transforms.py:16]: Also, I didn't want to install the whole package just for these 20 lines of code.
